src/complex/matrix.rs

```diff
@@ -0,0 +1,2290 @@
+    ///                                         Complex64::new(7f64, 7f64)],
+    ///                             2, 2, Row
+    ///     );
+    ///
+    ///     assert_eq!(y, y_row_map);
+    /// }
+    /// ```
+    fn row_map<F>(&self, f: F) -> ComplexMatrix
+    where
+        F: Fn(Vec<Complex<f64>>) -> Vec<Complex<f64>>,
+    {
+        let mut result = complex_matrix(
+            vec![Complex::zero(); self.row * self.col],
+            self.row,
+            self.col,
+            Shape::Row,
+        );
+
+        for i in 0..self.row {
+            result.subs_row(i, &f(self.row(i)));
+        }
+
+        result
+    }
+
+    fn col_mut_map<F>(&mut self, f: F)
+    where
+        F: Fn(Vec<Complex<f64>>) -> Vec<Complex<f64>>,
+    {
+        for i in 0..self.col {
+            unsafe {
+                let mut p = self.col_mut(i);
+                let fv = f(self.col(i));
+                for j in 0..p.len() {
+                    *p[j] = fv[j];
+                }
+            }
+        }
+    }
+
+    fn row_mut_map<F>(&mut self, f: F)
+    where
+        F: Fn(Vec<Complex<f64>>) -> Vec<Complex<f64>>,
+    {
+        for i in 0..self.col {
+            unsafe {
+                let mut p = self.row_mut(i);
+                let fv = f(self.row(i));
+                for j in 0..p.len() {
+                    *p[j] = fv[j];
+                }
+            }
+        }
+    }
+
+    fn reduce<F, T>(&self, init: T, f: F) -> Complex<f64>
+    where
+        F: Fn(Complex<f64>, Complex<f64>) -> Complex<f64>,
+        T: Into<Complex<f64>>,
+    {
+        self.data.iter().fold(init.into(), |x, y| f(x, *y))
+    }
+
+    fn zip_with<F>(&self, f: F, other: &ComplexMatrix) -> Self
+    where
+        F: Fn(Complex<f64>, Complex<f64>) -> Complex<f64>,
+    {
+        assert_eq!(self.data.len(), other.data.len());
+        let mut a = other.clone();
+        if self.shape != other.shape {
+            a = a.change_shape();
+        }
+        let result = self
+            .data
+            .iter()
+            .zip(a.data.iter())
+            .map(|(x, y)| f(*x, *y))
+            .collect::<Vec<Complex<f64>>>();
+        complex_matrix(result, self.row, self.col, self.shape)
+    }
+
+    fn col_reduce<F>(&self, f: F) -> Vec<Complex<f64>>
+    where
+        F: Fn(Vec<Complex<f64>>) -> Complex<f64>,
+    {
+        let mut v = vec![Complex::zero(); self.col];
+        for i in 0..self.col {
+            v[i] = f(self.col(i));
+        }
+        v
+    }
+
+    fn row_reduce<F>(&self, f: F) -> Vec<Complex<f64>>
+    where
+        F: Fn(Vec<Complex<f64>>) -> Complex<f64>,
+    {
+        let mut v = vec![Complex::zero(); self.row];
+        for i in 0..self.row {
+            v[i] = f(self.row(i));
+        }
+        v
+    }
+}
+
+// =============================================================================
+// Linear Algebra
+// =============================================================================
+
+/// Linear algebra trait
```

Comment:
Suggestion: Instead of commenting out unimplemented methods, consider using the `unimplemented!()` macro. This provides a clear indication that the functionality is planned but not yet implemented, and will panic if accidentally called. For example:

fn lu(&self) -> PQLU {
    unimplemented!("LU decomposition not yet implemented")
}

This approach maintains the method signatures in the trait, making it easier to track which methods still need implementation.

Reply:
In my opinion, I would rather remove the commented-out functions for now and add a ToDo note. The reason being, that to add the function signatures, the structs, enum etc needs to be defined/added. I am not a fan to adding such structs and enums before-hand and wait for it's use in the future.

For example,
```rust
fn lu(&self) -> PQLU {
unimplemented!("LU decomposition not yet implemented")
}
```
For this, I need to define PQLU in the code. There is no reason to define this now and wait for its usage sometime in the future. What do you think?

Reply:
I agree with your approach. It's a sensible way to keep the codebase clean and focused. Let's proceed with your suggestion.